pkg/controllers/podgroup/pg_controller.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2019 The Volcano Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podgroup
+
+import (
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	kbver "github.com/kubernetes-sigs/kube-batch/pkg/client/clientset/versioned"
+	kbinfoext "github.com/kubernetes-sigs/kube-batch/pkg/client/informers/externalversions"
+	kbinfo "github.com/kubernetes-sigs/kube-batch/pkg/client/informers/externalversions/scheduling/v1alpha1"
+	kblister "github.com/kubernetes-sigs/kube-batch/pkg/client/listers/scheduling/v1alpha1"
+
+	vkver "volcano.sh/volcano/pkg/client/clientset/versioned"
+)
+
+// Controller the Podgroup Controller type
+type Controller struct {
+	kubeClients *kubernetes.Clientset
+	vkClients   *vkver.Clientset
+	kbClients   *kbver.Clientset
+
+	podInformer     coreinformers.PodInformer
+	pgInformer      kbinfo.PodGroupInformer
+	sharedInformers informers.SharedInformerFactory
+
+	// A store of pods
+	podLister corelisters.PodLister
+	podSynced func() bool
+
+	// A store of podgroups
+	pgLister kblister.PodGroupLister
+	pgSynced func() bool
+
+	queue workqueue.RateLimitingInterface
+}
+
+// NewPodgroupController create new Podgroup Controller
+func NewPodgroupController(
+	kubeClient *kubernetes.Clientset,
+	kbClient *kbver.Clientset,
+	vkClient *vkver.Clientset,
+) *Controller {
+	cc := &Controller{
+		kubeClients: kubeClient,
+		vkClients:   vkClient,
+		kbClients:   kbClient,
+
+		queue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+
+	cc.sharedInformers = informers.NewSharedInformerFactory(cc.kubeClients, 0)
+	cc.podInformer = cc.sharedInformers.Core().V1().Pods()
+	cc.podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: cc.addPod,
```

Comment:
Why not watch job and create podgroup when necessary? there would be N times event triggered for this logic.

Reply:
maybe user have new crd job.haha.xxx, we can't catch.